Firestore/core/src/firebase/firestore/core/relation_filter.cc

```diff
@@ -41,48 +67,76 @@ const FieldPath& RelationFilter::field() const {
 
 bool RelationFilter::Matches(const model::Document& doc) const {
   if (field_.IsKeyFieldPath()) {
-    // TODO(rsgowman): Port this case
-    abort();
+    HARD_ASSERT(value_rhs_.type() == FieldValue::Type::Reference,
+                "Comparing on key, but filter value not a Reference.");
+    HARD_ASSERT(op_ != Filter::Operator::ArrayContains,
+                "arrayContains queries don't make sense on document keys.");
+    const auto& ref = value_rhs_.reference_value();
+    ComparisonResult comparison = doc.key().CompareTo(ref.key());
+    return MatchesComparison(comparison);
   } else {
```

Comment:
Optional: consider removing this `else` to streamline the code.

Reply:
The above if isn't really a guard condition so much as an alternative. Since it should never fall through, I prefer to keep the else. This ensures the cases are mutually exclusive.